### PR TITLE
feat(ndt_omp): add multigrid_ndt_omp for dynamic map loading 

### PR DESCRIPTION
**Pull request by kminoda**
See https://github.com/autowarefoundation/autoware.universe/pull/2339

### CODE REVIEW

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+      /** \brief Initializes voxel structure.
+       * \param[in] searchable flag if voxel structure is searchable, if true then kdtree is built
+       */
+      inline void
+      setInputCloudAndFilter (const PointCloudConstPtr &cloud, const std::string cloud_id, bool searchable = false)
+      {
+        searchable_ = searchable;
+        VoxelGridInfo voxel_grid_info;
+        applyFilter ((*cloud).makeShared(), cloud_id, voxel_grid_info); // TODO koji: yabai
```
Comment (KYabuuchi) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: It looks yabai copy.
I think we can avoid redundant copy by changing the argument type of the `applyFilter()`. Wouldn't that work?

`void applyFilter (const PointCloud`**Const**`Ptr &input, const std::string & cloud_id, VoxelGridInfo &voxel_grid_info) const;`
then, 
`applyFilter (cloud, cloud_id, voxel_grid_info);`

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    LeafID leaf_idx = getLeafID(cloud_id, input->points[cp], bbox);
+    Leaf& leaf = voxel_grid_info.leaves[leaf_idx];
+    updateLeaf(input->points[cp], centroid_size, leaf);
+  }
+
+  // Second pass: go over all leaves and compute centroids and covariance matrices
+  voxel_grid_info.voxel_centroids.points.reserve(voxel_grid_info.leaves.size ());
+  if (searchable_)
+    voxel_grid_info.leaf_indices.reserve(voxel_grid_info.leaves.size ());
+  int cp = 0;
```
Comment (KYabuuchi) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: `int cp = 0;` is not used.

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 1 above)
Comment (kminoda) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Thanks, I completely forgot this ToDo.
It seems that your proposal looks nice. I reflected the idea here: https://github.com/tier4/ndt_omp/pull/19/commits/bcf24a826ddf1a356cb21f5d76392819bc59344f

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        min_b_.setZero ();
+        max_b_.setZero ();
+        filter_name_ = "MultiVoxelGridCovariance";
+      }
+
+      /** \brief Initializes voxel structure.
+       * \param[in] searchable flag if voxel structure is searchable, if true then kdtree is built
+       */
+      inline void
+      setInputCloudAndFilter (const PointCloudConstPtr &cloud, const std::string cloud_id, bool searchable = false)
```
Comment (KYabuuchi) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: I think `const std::string & cloud_id` is better.  Copying strings can be slow.

Similarly, `removeCloud(const std::string cloud_id)` should be modified.

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,1197 @@
...
+		// 	target_cells_.getNeighborhoodAtPoint(x_trans_pt, neighborhood);
+		// 	break;
+		// default:
+		// case DIRECT7:
+		// 	target_cells_.getNeighborhoodAtPoint7(x_trans_pt, neighborhood);
+		// 	break;
+		// case DIRECT1:
+		// 	target_cells_.getNeighborhoodAtPoint1(x_trans_pt, neighborhood);
+		// 	break;
+		// }
```
Comment (KYabuuchi) [inline include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: These comments are no longer relevant, but you won't delete it?

Comment (KYabuuchi) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: I am mentioning this section.
image

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp (hunk 2 above)
Comment (kminoda) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp_impl.hpp]: Resolved: https://github.com/tier4/ndt_omp/pull/19/commits/24e52cd7cf998a6a44e36aad6fb8b9ec4150111a

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h (hunk 3 above)
Comment (kminoda) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Resolved: https://github.com/tier4/ndt_omp/pull/19/commits/24e52cd7cf998a6a44e36aad6fb8b9ec4150111a

File: include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp (hunk 4 above)
Comment (kminoda) [reply include/multigrid_pclomp/multigrid_ndt_omp_impl.hpp]: Resolved: https://github.com/tier4/ndt_omp/pull/19/commits/24e52cd7cf998a6a44e36aad6fb8b9ec4150111a & https://github.com/tier4/ndt_omp/pull/19/commits/b29f6c41403a9ca593c46f488d9eb02691a76ff9

File: include/multigrid_pclomp/multigrid_ndt_omp.h [hunk 5]
```diff
@@ -0,0 +1,660 @@
...
+		int max_iterations;
+		// pclomp::NeighborSearchMethod search_method;
+		int num_threads;
+		float regularization_scale_factor;
+	};
+
+	/** \brief A 3D Normal Distribution Transform registration implementation for point cloud data.
+	  * \note For more information please see
+	  * <b>Magnusson, M. (2009). The Three-Dimensional Normal-Distributions Transform —
+	  * an Efficient Representation for pcl::Registration, Surface Analysis, and Loop Detection.
```
Comment (YamatoAndo) [inline include/multigrid_pclomp/multigrid_ndt_omp.h]: Probably an unintended change.
```suggestion
	  * an Efficient Representation for Registration, Surface Analysis, and Loop Detection.
```

Comment (kminoda) [reply include/multigrid_pclomp/multigrid_ndt_omp.h]: Resolved: https://github.com/tier4/ndt_omp/pull/19/commits/ca5e438585d50f27e635bbc343aaddbcf8288c41

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+
+      /** \brief Const pointer to MultiVoxelGridCovariance leaf structure */
+      typedef const Leaf* LeafConstPtr;
+
+      typedef std::map<LeafID, Leaf> Map;
+
+      struct VoxelGridInfo {
+        PointCloud voxel_centroids;
+        Map leaves;
+        std::vector<LeafID> leaf_indices;
```
Comment (YamatoAndo) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: `leaves` includes information on `voxel_centroids` and `leaf_indices`.

Comment (kminoda) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Thank you, you are right.
I removed the VoxelGridInfo struct itself in the latest code. Please check the latest three commits.
https://github.com/tier4/ndt_omp/pull/19/commits/77a104015afc82db656443aeed781168b2a0265e
https://github.com/tier4/ndt_omp/pull/19/commits/c632fc80b9f9f04851900c4fa85d66fe5bf3717b
https://github.com/tier4/ndt_omp/pull/19/commits/c63423b3fd5d8089a99c649e392d4825c6c1a6d4

File: include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+      void updateLeaf (const PointT &point, const int &centroid_size, Leaf &leaf) const;
+
+      void computeLeafParams (const Eigen::Vector3d &pt_sum,
+        Eigen::SelfAdjointEigenSolver<Eigen::Matrix3d> &eigensolver,
+        Leaf &leaf) const;
+
+      LeafID getLeafID (const std::string &grid_id, const PointT &point, const BoundingBox &bbox) const;
+
+      /** \brief Flag to determine if voxel structure is searchable. */
+      bool searchable_;
```
Comment (KYabuuchi) [inline include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: `searchable_` does not have any rolls (It must always be true. ) so you can remove it completely.

Comment (kminoda) [reply include/multigrid_pclomp/multi_voxel_grid_covariance_omp.h]: Thanks, I removed the variable: https://github.com/tier4/ndt_omp/pull/19/commits/d0f7829ce67968b245b1f13a5745331ec157f542

**Comment by YamatoAndo** [comment]
@kminoda There are many modifications from the original code.
So please check if you need to add your own license to the header files.

**Comment by kminoda** [comment]
@YamatoAndo Based on the fact that [reeds_shepp.cpp](https://github.com/autowarefoundation/autoware.universe/blob/1bab84e75e6a6b750b5781212b04702b7646616f/planning/freespace_planning_algorithms/src/reeds_shepp.cpp) claims dual license (original + TIER IV), which is heavily based on [pyReedsShepp](https://github.com/ghliu/pyReedsShepp/blob/master/reeds_shepp/src/reeds_shepp.cpp), I will add TIER IV license on the additional scripts.

**Review by KYabuuchi — APPROVED** [review]
I have verified this on a large map dataset and NDT seems to work correctly :+1:

**Pull request merged**